src/psyclone/dynamo0p3.py

```diff
@@ -1684,24 +1701,33 @@ def _stub_declarations(self, parent):
         if self._kernel.cma_operation not in ["apply", "matrix-matrix"]:
             parent.add(DeclGen(parent, datatype="integer",
                                kind=api_config.default_kind["integer"],
-                               intent="in", entity_decls=[self._nlayers_name]))
+                               intent="in",
+                               entity_decls=list(self._nlayers_names)))
 
     def initialise(self, parent):
         '''
-        Look-up the number of vertical layers in the mesh in the PSy layer.
+        Look-up the number of vertical layers in the mesh for each user-
+        supplied kernel that operates on cell columns.
 
         :param parent: the f2pygen node representing the PSy-layer routine.
         :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
 
         '''
-        if not self._dofs_only:
-            parent.add(CommentGen(parent, ""))
-            parent.add(CommentGen(parent, " Initialise number of layers"))
-            parent.add(CommentGen(parent, ""))
+        if not self._nlayers_names:
+            return
+
+        parent.add(CommentGen(parent, ""))
+        parent.add(CommentGen(parent, " Initialise number of layers"))
+        parent.add(CommentGen(parent, ""))
+        # Sort for test reproducibility
+        sorted_names = list(self._nlayers_names.keys())
```

Comment:
Again, it is (apparently :) ) certain that _nlaers_names is a dict here.

Reply:
I've tidied this up now.